### PR TITLE
WW-5697 Restrict the indexed-access fast path in XWorkMethodAccessor to real indexed properties

### PR DESCRIPTION
**Pull request by lukaszlenart**
Fixes [WW-5697](https://issues.apache.org/jira/browse/WW-5697)

## Problem

`XWorkMethodAccessor.callMethod(...)` skipped the `denyMethodExecution` check for any method whose name began with `get` and took one argument, or `set` and took two:

```java
//HACK - we pass indexed method access i.e. setXXX(A,B) pattern
if ((objects.length == 2 && string.startsWith("set")) || (objects.length == 1 && string.startsWith("get"))) {
```

That is a name prefix plus an argument count, not a property check. An ordinary method such as `getSomething(String)` is not a JavaBeans property, but it matches, so it was executed during parameter binding with the argument taken from the parameter name — a parameter name of `getSomething('value').property` calls `getSomething("value")`.

The flag the branch consults, `DENY_INDEXED_ACCESS_EXECUTION`, is never written anywhere in main source, so `exec` is always `null` and the fast path is unconditional.

## Change

The fast path now applies only where the call really is the indexed accessor of a property on the target type. The property name alone is not enough to decide that, which review of the first cut showed the hard way:

- a class declaring the indexed pair `getItem(int)` / `setItem(int, String)` may **also** declare an unrelated `getItem(String)` overload, and a one-argument call dispatches to *that* overload — the argument types choose the method, and the caller chooses the arguments;
- the check has to know its direction, or a read-only `getItem(int)` legitimises an unrelated two-argument `setItem(String, String)`.

So the descriptor's own indexed accessor must be the method that will actually run: same name, same direction (`getIndexedReadMethod` / `getIndexedWriteMethod`, for both `IndexedPropertyDescriptor` and OGNL's `ObjectIndexedPropertyDescriptor`), and no same-arity overload for the dispatcher to prefer instead. Everything else falls through to the `denyMethodExecution` check.

That last clause counts candidates by *signature*, not by `Method`. `OgnlRuntime.getMethods` reports an overridden method and the method overriding it separately, so counting `Method` objects mistook an override for an overload and denied an ordinary bean shape — a base class declaring the indexed pair, a subclass refining the accessor. Two methods related by an override share a parameter list and so are not a choice the dispatcher makes; only one implementation can ever run. Distinct parameter lists of the same arity are the real overloads, and still deny.

The deny check is hoisted ahead of the indexed-property block, which it now guards. The two are equivalent — with execution permitted, both paths ended in the same `callMethodWithDebugInfo` call — but this way the introspection is skipped entirely on the common path, and the block reads as the exception it is.

`DENY_INDEXED_ACCESS_EXECUTION` is public API, so it is deprecated here rather than deleted; removal in 8.0.0 is tracked as [WW-5699](https://issues.apache.org/jira/browse/WW-5699). It is still honoured, so the framework's own read of it carries `@SuppressWarnings("removal")` rather than warning on every core build.

## Scope of the behaviour change

Confined to parameter binding. The fast path only *matters* when the deny flag is set, and that flag is set only by `ParametersInterceptor`, `AliasInterceptor` and `StaticParametersInterceptor` — with it unset, such a call already fell through to the check below and executed. JSP and tag rendering are unaffected.

Two narrowings worth calling out for the migration guide, both silent (debug-level) when they bite, and both only while method execution is denied:

- **An unpaired object-indexed getter is no longer exempt.** `String getKeyed(String)` with no matching `setKeyed(String, String)` reports `INDEXED_PROPERTY_NONE`, because OGNL's `findObjectIndexedPropertyDescriptors` requires an exact pair with matching key and value types. The same applies when the types do not line up, e.g. `String getKeyed(String)` against `void setKeyed(String, Object)`. OGNL itself refuses `bean.keyed['k']` for such beans, so this is consistent, but it is a change.
- **Bare `get(...)` / `set(...)` are no longer exempt.** `"get".startsWith("get")` matched the old fast path, so `myMap.get('k').field` bound under `denyMethodExecution`. There is no property name left once the prefix is stripped, so it cannot be an indexed accessor.

## Tests

`XWorkMethodAccessorTest` is new. Among the behaviours it covers:

- an argument-taking getter that is not an indexed property is not executed while method execution is denied
- an **overload** of a genuine indexed accessor is not executed while denied
- an unrelated two-argument setter named after a **read-only** indexed property is not executed while denied
- object-indexed and int-indexed accessors still execute while denied
- a bare `get(String)` is not executed while denied
- with the deny flag unset, an argument-taking getter still executes, as before
- an indexed accessor **overridden** in a subclass still executes while denied

Note that a `getItem(int)` / `setItem(int, String)` pair reports as `INDEXED_PROPERTY_OBJECT`, not `_INT`: `findObjectIndexedPropertyDescriptors` overwrites the `java.beans` descriptor whenever it finds a matching pair. A **read-only** int-indexed getter is the only shape that reaches the `_INT` branch, so there is a test for it specifically.

Mutation-checked both ways: restoring the old name-only predicate fails exactly the two overload/direction tests, and forcing the predicate to always reject fails exactly the three "still works" tests. None of them pass vacuously. The override test was written first and watched fail against the `Method`-counting predicate, for the right reason — the accessor was denied and the call returned null.

Full `core` suite green: 3275 tests, 0 failures, 0 errors.

## Related

[WW-5698](https://issues.apache.org/jira/browse/WW-5698) covers a separate issue found alongside this one — the `ModelDriven` exemption in `StrutsParameterAuthorizer` also exempting the action's own members. Different cause, different fix, not addressed here. An end-to-end `ModelDriven` binding test was deliberately left out of this PR because it would couple these tests to the exemption WW-5698 is expected to change.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: core/src/main/java/org/apache/struts2/util/reflection/ReflectionContextState.java [hunk 1]
```diff
@@ -38,6 +38,13 @@ public class ReflectionContextState {
 	public static final String FULL_PROPERTY_PATH = "current.property.path"; // TODO: Probably a bug
 	public static final String CREATE_NULL_OBJECTS = "xwork.NullHandler.createNullObjects";
 	public static final String DENY_METHOD_EXECUTION = "xwork.MethodAccessor.denyMethodExecution";
+	/**
+	 * @deprecated since 7.4.0, no replacement. Nothing in the framework has ever set this key, so it has
+	 * never had any effect. Indexed property access is now identified by inspecting the target type rather
+	 * than by trusting a method name prefix, which leaves this flag with nothing to guard. Scheduled for
+	 * removal in 8.0.0 by WW-5699.
```
[the remark this answers is not in the document]
Comment (lukaszlenart) [reply core/src/main/java/org/apache/struts2/util/reflection/ReflectionContextState.java]: Reworded in the current head. The javadoc now says Struts core never sets the key, and states what does still hold: application or plugin code that sets it itself continues to suppress the fast path, which is why the constant is deprecated rather than removed outright.

That behaviour also has tests now, both ways — the key set, and the key set to `false`.

File: core/src/main/java/org/apache/struts2/ognl/accessor/XWorkMethodAccessor.java [hunk 2]
```diff
@@ -94,6 +99,23 @@ public Object callMethod(OgnlContext context, Object object, String string, Obje
...
+            return false;
+        }
+        String propertyName = Introspector.decapitalize(methodName.substring(3));
+        try {
+            return OgnlRuntime.getIndexedPropertyType(object.getClass(), propertyName) != OgnlRuntime.INDEXED_PROPERTY_NONE;
+        } catch (OgnlException e) {
+            LOG.debug("Could not determine whether [{}] is an indexed property of [{}]", propertyName, object.getClass(), e);
+            return false;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (lukaszlenart) [reply core/src/main/java/org/apache/struts2/ognl/accessor/XWorkMethodAccessor.java]: Not adding a cache — the expensive half is already cached by OGNL.

`OgnlRuntime.getPropertyDescriptor` goes through `getPropertyDescriptors`, which reads `OgnlCache.propertyDescriptorCache`, a `ClassCache<Map<String, PropertyDescriptor>>` (`OgnlCache.java:129`), so `Introspector.getBeanInfo` runs once per class rather than once per invocation. `getMethods(Class, String, boolean)` is cached the same way. A second cache in front of those would duplicate them and add another class-keyed static map, which is the shape WW-5537 was about.

What is left per call is a `startsWith`, a `substring` and a `decapitalize`. The check has also moved behind the deny test in the current head, so none of it runs unless method execution is actually denied.

File: core/src/main/java/org/apache/struts2/ognl/accessor/XWorkMethodAccessor.java [hunk 3]
```diff
@@ -77,21 +83,81 @@ public Object callMethod(OgnlContext context, Object object, String string, Obje
...
+            return callMethodWithDebugInfo(context, object, string, objects);
         }
-        boolean e = ReflectionContextState.isDenyMethodExecution(context);
 
-        if (!e) {
+        //Method execution is denied. Indexed property access, i.e. the getXXX(A) / setXXX(A,B) pattern, is
+        //the one exception, because reading a['k'] must keep working during parameter binding. It is
+        //restricted to calls which really are the indexed accessor of a property on the target type: a name
+        //prefix and an argument count alone would let any method be called while execution is denied.
+        if (isIndexedPropertyAccessor(object, string, objects)
```
[the remark this answers is not in the document]
Comment (lukaszlenart) [reply core/src/main/java/org/apache/struts2/ognl/accessor/XWorkMethodAccessor.java:94]: Confirmed — this was checked before the PR was opened, and again against SECURITY.md.

WW-5697 was triaged as an ordinary bug rather than a vulnerability: the control defect is ours, but the impact is application-owned, so it is handled publicly with no CVE and no security framing. The dead `DENY_INDEXED_ACCESS_EXECUTION` constant is removed separately under WW-5699, targeted at 8.0.0.
